File: src/main/java/de/arrobait/antlers/editor/structure/AntlersStructuredViewElement_.java
package de.arrobait.antlers.editor.structure;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.structureView.StructureViewTreeElement;
import com.intellij.ide.util.treeView.smartTree.SortableTreeElement;
import com.intellij.ide.util.treeView.smartTree.TreeElement;
import com.intellij.navigation.ItemPresentation;
import com.intellij.psi.NavigatablePsiElement;
import com.intellij.psi.util.PsiTreeUtil;
import de.arrobait.antlers.psi.AntlersFile;
import de.arrobait.antlers.psi.AntlersTine;
import de.arrobait.antlers.psi.impl.AntlersTineImpl;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

//public class AntlersStructuredViewElement implements StructureViewTreeElement, SortableTreeElement {
public class AntlersStructuredViewElement_ implements StructureViewTreeElement, SortableTreeElement {

    private final NavigatablePsiElement myElement;

    public AntlersStructuredViewElement_(NavigatablePsiElement myElement) {
        this.myElement = myElement;
    }

    @Override
    public Object getValue() {
        return myElement;
    }

    @Override
    public @NotNull String getAlphaSortKey() {
        String name = myElement.getName();

        return name != null ? name : "";
    }

    @Override
    public @NotNull ItemPresentation getPresentation() {
        ItemPresentation presentation = myElement.getPresentation();

        return presentation != null ? presentation : new PresentationData();
    }

    @Override
    public TreeElement @NotNull [] getChildren() {
        if (myElement instanceof AntlersFile) {
            List<AntlersTine> tines = PsiTreeUtil.getChildrenOfTypeAsList(myElement, AntlersTine.class);
            List<TreeElement> treeElements = new ArrayList<>(tines.size());
            for (AntlersTine tine : tines) {
                treeElements.add(new AntlersStructuredViewElement_((AntlersTineImpl) tine));
            }
            return treeElements.toArray(new TreeElement[0]);

        }

        return EMPTY_ARRAY;
    }

    @Override
    public void navigate(boolean requestFocus) {

    }

    @Override
    public boolean canNavigate() {
        return false;
    }

    @Override
    public boolean canNavigateToSource() {
        return false;
    }
}

File: src/main/java/de/arrobait/antlers/editor/structure/AntlersStructuredViewElement.kt
package de.arrobait.antlers.editor.structure

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.structureView.StructureViewTreeElement
import com.intellij.ide.structureView.impl.common.PsiTreeElementBase
import com.intellij.navigation.ItemPresentation
import com.intellij.psi.NavigatablePsiElement
import com.intellij.psi.PsiElement
import com.intellij.util.ReflectionUtil
import de.arrobait.antlers.psi.AntlersPsiElement
import de.arrobait.antlers.psi.AntlersTine

class AntlersStructuredViewElement(private val element: NavigatablePsiElement) : PsiTreeElementBase<NavigatablePsiElement>(element) {
    companion object {
        fun getStructuredViewTreeElements(psiElement: NavigatablePsiElement): MutableList<StructureViewTreeElement> {
            val children: MutableList<StructureViewTreeElement> = arrayListOf()
            for (childElement: PsiElement in psiElement.children) {
                if (childElement !is AntlersPsiElement) {
                    continue
                }

                if (childElement is AntlersTine) {
                    children.addAll(
                        AntlersStructuredViewElement(childElement as AntlersPsiElement).childrenBase
                    )
                }

                for (suitableClass in AntlersStructureViewModel.ourSuitableClasses) {
                    if (ReflectionUtil.isAssignable(suitableClass, childElement.javaClass)) {
                        children.add(AntlersStructuredViewElement(childElement))
                        break
                    }
                }
            }

            return children
        }
    }

    override fun getChildrenBase(): MutableCollection<StructureViewTreeElement> {
        return getStructuredViewTreeElements(element)
    }

    override fun getPresentableText(): String? = element.name

    override fun getPresentation(): ItemPresentation {
        return element.presentation ?: PresentationData()
    }

    override fun navigate(requestFocus: Boolean) = element.navigate(requestFocus)

    override fun canNavigate(): Boolean = element.canNavigate()

    override fun canNavigateToSource(): Boolean = element.canNavigateToSource()
}
